app/dao/notifications_dao.py

```diff
@@ -806,3 +806,17 @@ def get_service_ids_that_have_notifications_from_before_timestamp(notification_t
             Notification.created_at < timestamp
         ).distinct()
     }
+
+
+def get_service_ids_with_notifications_on_date(notification_type, date):
+    return {
+        row.service_id
+        for row in db.session.query(
+            Notification.service_id
+        ).filter(
+            Notification.notification_type == notification_type,
+            # using >= + < is much more efficient than date(created_at)
+            Notification.created_at >= date,
+            Notification.created_at < date + timedelta(days=1)
```

Comment:
hmmm, alarm bells are ringing here. could this lead to weird bugs on the BST/GMT changeover dates? 

in fact now i see that we're already doing similar calculations in the outer loop just subtracting `days=1`.

I suspect that for example if you run this just after the late march switch, it'll keep the same time (11:00pm UTC) for each day, even if those days are before the DST switch and should be from 00:00. That could lead to a bit of double counting. 😬

I think (and this is very much something i've not tested) that we should do the calculation with `date` objects and then call `get_london_midnight_in_utc` on that date object each time we want to compare to a `created_at` timestamp

Reply:
Great spot @leohemsted !

It took me a little while to get my head around this, so I'll write some notes here with my understanding and then look at addressing your comment.

## What do we want?

- The day I think I'm running the task on is the day it actually runs on. This is always the case as the BST changeover is either at 1AM (forwards) or 2AM (backwards), so we'll always lock on to the correct day.

- The set of notifications for a day should match the timezone we were in when they were created e.g. if we think it was created late on Monday for us, then it should be attributed to Monday. This is the case for [the actual query](https://github.com/alphagov/notifications-api/blob/de3e0eeb954f6eacc33fc5d96683f30f94c96bea/app/dao/fact_notification_status_dao.py#L39) but, as you point out here, it's not the case for the pre-check to get the relevant service IDs.

This wouldn't lead to double counting but could lead to under counting ("no counting") if a service only sent some notifications between midnight and 1AM on a single, isolated day (once we move to BST):

- We look for 00:00 - 00:00 in the DB (UTC)
- But notifications really span from (in UTC):

   - 26/03 00:00 - 27/03 00:00 (for 26/03, GMT)
   - 27/03 00:00 - 27/03 23:00 (for 27/03, GMT/BST)
   - 27/03 23:00 - 28/03 23:00 (for 28/03, BST)

- So a notification sent at 28/03 00:30 (BST) - 27/03 23:30 (UTC) - would incorrectly trigger an aggregation for the associated service on the 27th, instead of the 28th. A bit of an edge case, but worth fixing 👍.

## How do we get it?

Do what you said ;-).

At the same time, we can simplify [the actual query](https://github.com/alphagov/notifications-api/blob/de3e0eeb954f6eacc33fc5d96683f30f94c96bea/app/dao/fact_notification_status_dao.py#L39) to use the same approach and put some tests around both functions to make the "what we want" more explicit (and correct).

Reply:
> At the same time, we can simplify [the actual query](https://github.com/alphagov/notifications-api/blob/de3e0eeb954f6eacc33fc5d96683f30f94c96bea/app/dao/fact_notification_status_dao.py#L39)

Addressed in https://github.com/alphagov/notifications-api/pull/3437/commits/f937d833b0aa3e0f00ac77603424e5570cd1657b

> Do what you said ;-).

Addressed in https://github.com/alphagov/notifications-api/pull/3437/commits/54163611e7f1f798b4dae99cfab0644d63c3763c.